GH Action: merge staging(-next) periodically

FRidh opened the pull request:
Automate the merging of `master` -> `staging-next` -> `staging`.

Our main development branch is `master`. Large rebuilds go to `staging`.
Periodically, `staging` is merged into `staging-next` for stabilization.
When considered sufficiently stable, `staging-next` is merged into
`master`.

As changes arrive on these branches, it is important that they're all
updated regularly with eachothers changes. This commit automates that
part.

https://github.com/NixOS/nixpkgs/issues/104594

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

FRidh committed:
GH Action: merge staging(-next) periodically

Automate the merging of `master` -> `staging-next` -> `staging`.

Our main development branch is `master`. Large rebuilds go to `staging`.
Periodically, `staging` is merged into `staging-next` for stabilization.
When considered sufficiently stable, `staging-next` is merged into
`master`.

As changes arrive on these branches, it is important that they're all
updated regularly with eachothers changes. This commit automates that
part.

SuperSandro2000 commented:
How does this handle merge conflicts and conflicting/undoing changes? Rebasing is also not an option because it could easily create merge conflicts.

FRidh commented:
If there is a merge conflict I would *expect* this action to fail, resulting in a notification (but for who?).

This probably should be tested on another repo first.

lukegb commented:
You probably want to restrict this to run on the main NixOS/nixpkgs repo (e.g. `if: github.repository == ‘NixOS/nixpkgs’`)

FRidh commented:
Configured it on my fork https://github.com/FRidh/nixpkgs/commit/75fe13617a0b3fa9b03dd230429bc6a21c5b2c68.

zowoq commented on .github/workflows/merge-staging.yml:
```diff
@@ -0,0 +1,30 @@
+    if: github.repository == 'NixOS/nixpkgs'
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@master
```
```suggestion
      - uses: actions/checkout@v2
```
Upstream seems to recommend pinning this in their examples. https://github.com/actions/checkout

FRidh commented:
Example of a merge conflict https://github.com/FRidh/nixpkgs/runs/1467125543?check_suite_focus=true.

Mic92 commented on .github/workflows/merge-staging.yml:
```diff
@@ -0,0 +1,30 @@
+  schedule:
+    # * is a special character in YAML so you have to quote this string
+    # Merge every 6 hours
+    - cron:  '* */6 * * *'
```
How about publishing merge conflicts notifications on IRC?

FRidh replied:
That's a good idea. Notifying on IRC could be done with e.g. https://github.com/marketplace/actions/irc-message-action

FRidh committed:
GH Action: merge staging(-next): comment on failure

Comment on the designated issue in case of a failure.

Whoever wants to help out resolving merge conflicts can simply follow
this issue.

FRidh commented:
I added a step that would notify on an issue (currently this one) when the workflow fails.

Mic92 commented:
sounds good.

lukegb commented:
> I added a step that would notify on an issue (currently this one) when the workflow fails.

Hooray, it worked :^)

Mic92 commented:
> An automatic merge [failed](https://github.com/NixOS/nixpkgs/actions/runs/403256027).

fixed.

FRidh commented:
> An automatic merge [failed](https://github.com/NixOS/nixpkgs/actions/runs/415367940).

Fixed with a717b55392ffd9c558b6135662c894fa69d4d82d.

primeos commented:
Any particular reason why the recent automatic merge failures are being ignored?
(Note: Please don't take this as a complaint. I'm just trying to understand our new workflow and wondering if this is intentional (e.g. during staging-next PRs) or if it's due to time constraints, notification issues, etc.).

jtojnar commented:
Probably no one got around to resolving them, will do that now.

jtojnar commented:
Done in f19eb635b45.

mweinelt commented:
How could we reimplement that logic efficiently for release-21.05/staging-21.05/staging-next-21.05?

Mic92 commented:
it would be great if we could convert this to a prometheus metric rather than an issue to see if resolves itself.

chrissyjai commented:
why am i suscribed to this

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: github-actions[bot] ***@***.***>
Sent: Sunday, March 27, 2022 7:02:47 PM
To: NixOS/nixpkgs ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [NixOS/nixpkgs] GH Action: merge staging(-next) periodically (#105153)


Periodic merge from master into staging-next has failed<https://github.com/NixOS/nixpkgs/actions/runs/2049374837>.

—
Reply to this email directly, view it on GitHub<https://github.com/NixOS/nixpkgs/pull/105153#issuecomment-1080049408>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AUFPY6M5AI3GWZIXUJ2VD4LVCDZKPANCNFSM4UE727KQ>.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

jtojnar commented:
@chrissyjai Maybe https://github.com/NixOS/nixpkgs/issues/165471?
